jetty-client/src/main/java/org/eclipse/jetty/client/util/AbstractRequestContent.java

```diff
@@ -0,0 +1,250 @@
+        return newSubscription;
+    }
+
+    protected abstract Subscription newSubscription(Consumer consumer, boolean emitInitialContent, Throwable failure);
+
+    @Override
+    public void fail(Throwable failure)
+    {
+        Subscription subscription = null;
+        try (AutoLock ignored = lock.lock())
+        {
+            if (this.failure == null)
+            {
+                this.failure = failure;
+                subscription = this.subscription;
+            }
+        }
+        if (subscription != null)
+            subscription.fail(failure);
+    }
+
+    public abstract class AbstractSubscription implements Subscription
+    {
+        private final Consumer consumer;
+        private final boolean emitInitialContent;
+        private Throwable failure;
+        private int demand;
+        private boolean stalled;
+        private boolean committed;
+
+        public AbstractSubscription(Consumer consumer, boolean emitInitialContent, Throwable failure)
+        {
+            this.consumer = consumer;
+            this.emitInitialContent = emitInitialContent;
+            this.failure = failure;
+            this.stalled = true;
+        }
+
+        @Override
+        public void demand()
+        {
+            boolean produce;
+            try (AutoLock ignored = lock.lock())
+            {
+                ++demand;
+                produce = stalled;
+                if (stalled)
+                    stalled = false;
+            }
+            if (LOG.isDebugEnabled())
+                LOG.debug("Content demand, producing {} for {}", produce, this);
+            if (produce)
+                produce();
+        }
+
+        private void produce()
+        {
+            while (true)
+            {
+                Throwable failure;
+                boolean committed;
+                try (AutoLock ignored = lock.lock())
+                {
+                    failure = this.failure;
+                    committed = this.committed;
+                }
+                if (failure != null)
+                {
+                    notifyFailure(failure);
+                    return;
+                }
+
+                if (committed || emitInitialContent)
+                {
+                    try
+                    {
+                        if (!produceContent(this::processContent))
+                            return;
+                    }
+                    catch (Throwable x)
+                    {
+                        // Fail and loop around to notify the failure.
+                        fail(x);
+                    }
+                }
+                else
+                {
+                    if (!processContent(BufferUtil.EMPTY_BUFFER, false, Callback.NOOP))
+                        return;
+                }
+            }
+        }
+
+        protected abstract boolean produceContent(Producer producer) throws Exception;
```

Comment:
what is the implementation mean to do?  call notifyContent?

Reply:
Will add javadocs.

The implementation is supposed to produce some content and give it to the `Producer` argument, or throw an exception if it cannot produce content.